Mathlib/GroupTheory/CosetCover.lean

```diff
@@ -0,0 +1,397 @@
+
+/-- Let the group `G` be the union of finitely many left cosets `g i • H i`. Then the
+sum of the inverses of the indexes of the subgroups `H i` is greater than or equal to 1. -/
+@[to_additive one_le_sum_inv_index_of_leftCoset_cover]
+theorem one_le_sum_inv_index_of_leftCoset_cover :
+    1 ≤ ∑ i ∈ s, ((H i).index : ℚ)⁻¹ :=
+  have := Classical.decPred (FiniteIndex : Subgroup G → Prop)
+  (leftCoset_cover_filter_FiniteIndex_aux hcovers).2.1
+
+/-- Let the group `G` be the union of finitely many left cosets `g i • H i`.
+If the sum of the inverses of the indexes of the subgroups `H i` is equal to 1,
+then the cosets of the subgroups of finite index are pairwise disjoint -/
+@[to_additive]
+theorem pairwiseDisjoint_leftCoset_cover_of_sum_inv_index_eq_one
+    [DecidablePred (FiniteIndex : Subgroup G → Prop)] :
+    ∑ i ∈ s, ((H i).index : ℚ)⁻¹ = 1 →
+      Set.PairwiseDisjoint (s.filter (fun i => (H i).FiniteIndex))
+        (fun i ↦ g i • (H i : Set G)) :=
+  (leftCoset_cover_filter_FiniteIndex_aux hcovers).2.2
+
+/-- B. H. Neumann Lemma :
+If a finite family of cosets of subgroups covers the group, then at least one
+of these subgroups has index not exceeding the number of cosets. -/
+@[to_additive]
+theorem exists_index_le_card_of_leftCoset_cover :
+    ∃ i ∈ s, (H i).FiniteIndex ∧ (H i).index ≤ s.card := by
+  classical
+  by_contra! h
+  apply (one_le_sum_inv_index_of_leftCoset_cover hcovers).not_lt
+  by_cases hs : s = ∅
+  · simp only [hs, Finset.sum_empty, show (0 : ℚ) < 1 from rfl]
+  rw [← ne_eq, ← Finset.nonempty_iff_ne_empty] at hs
+  have hs' : 0 < s.card := Nat.pos_of_ne_zero (Finset.card_ne_zero.mpr hs)
+  have hlt : ∀ i ∈ s, ((H i).index : ℚ)⁻¹ < (s.card : ℚ)⁻¹ := fun i hi ↦ by
+    cases (H i).index.eq_zero_or_pos with
+    | inl hindex =>
+      rwa [hindex, Nat.cast_zero, inv_zero, inv_pos, Nat.cast_pos]
+    | inr hindex =>
+      exact inv_lt_inv_of_lt (by exact_mod_cast hs') (by exact_mod_cast h i hi ⟨hindex.ne'⟩)
+  apply (Finset.sum_lt_sum_of_nonempty hs hlt).trans_eq
+  rw [Finset.sum_const, nsmul_eq_mul,
+    mul_inv_eq_iff_eq_mul₀ (Nat.cast_ne_zero.mpr hs'.ne'), one_mul]
+
+end
+
+end Subgroup
+
+section Submodule
+
+variable {R M ι : Type*} [Ring R] [AddCommGroup M] [Module R M]
+    {p : ι → Submodule R M} {s : Finset ι}
+    (hcovers : ⋃ i ∈ s, (p i : Set M) = Set.univ)
+
+theorem Submodule.exists_finiteIndex_of_cover :
+    ∃ k ∈ s, (p k).toAddSubgroup.FiniteIndex :=
+  have hcovers' : ⋃ i ∈ s, (0 : M) +ᵥ ((p i).toAddSubgroup : Set M) = Set.univ := by
+    simpa only [zero_vadd] using hcovers
+  AddSubgroup.exists_finiteIndex_of_leftCoset_cover hcovers'
+
+end Submodule
+
+section Subspace
+
+variable {k E ι : Type*} [DivisionRing k] [Infinite k] [AddCommGroup E] [Module k E]
+    {s : Finset (Subspace k E)}
+
+/- A vector space over an infinite field cannot be a finite union of proper subspaces. -/
+theorem Subspace.biUnion_ne_univ_of_ne_top (hs : ∀ p ∈ s, p ≠ ⊤) :
```

Comment:
Can this be generalized to free module over some class of rings?

Reply:
It's false for Z² as a Z-module, by taking the pre images of the three mod-2 lines. Maybe if all residue fields are infinite...

Reply:
OK, it's complicated. It doesn't matter.

Reply:
It's a good question, obviously, and in some sense, Neumann's lemma gives some answers. If you have a finitely generated module, then every submodule is contained in a maximal one whose co-quotient is a residue field, hence is infinite if all residue fields are infinite.

Reply:
But I have to admit that my way of proving the case of vector spaces is not via Neumann's lemma but via polynomials (over an infinite field, a nonzero polynomial does not vanish identically).